### ISSUE TITLE
Preprocessing of contentFiles can corrupt binary files

### ISSUE THREAD

**Issue by nguerrera** [dotnet/sdk#372]
We call `NuGet.CommonPreprocessor.Preprocess(Stream, Func<string, string>)` which returns a `String` and then we do File.WriteAllText.

I found this because @harshjain2 hit an issue with dlls in contentFiles, which is dubious, but it strikes me that we could also have other binary content such as a jpeg or whatever. 

Question: Is it it by design that everything in contentFiles/ in a package must be text? Where does non-text content go?

* If it is by design, then we need to have a better failure mode than possible silent corruption or crash.
* If it is not by design, then we need to make sure that binary files can pass through safely somehow.

@rrelyea for nuget question

**Comment by nguerrera** [comment]
There is also a perf concern if binaries go through the preprocessor even if it were changed to ensure no corruption of binaries.

**Comment by srivatsn** [comment]
@rrelyea should this move to the nuget repo?

**Comment by srivatsn** [comment]
Moved to https://github.com/NuGet/Home/issues/4138

**Comment by nguerrera** [comment]
See https://github.com/NuGet/Home/issues/4138#issuecomment-26937033, we're only supposed to be preprocessing *.pp.

**Comment by natidea** [comment]
[This check](https://github.com/dotnet/sdk/blob/aca5d1556f7c9b3e2182a38844d66be663ff26fb/src/Tasks/Microsoft.NET.Build.Tasks/ProduceContentAssets.cs#L218) is supposed to prevent this from happening (borrowed from the old [ResolveNuGetPackageAssets](https://github.com/NuGet/NuGet.BuildTasks/blob/4a57a0ef4bee31f122d50a39fe3dddd4ca03fb07/src/Microsoft.NuGet.Build.Tasks/ResolveNuGetPackageAssets.cs#L623) task):

```csharp
string ppOutputPath = contentFile.GetMetadata("ppOutputPath");
if (ppOutputPath != null)
{
    // process
}
```
Only content that incorrectly set the `ppOutputPath` metadata when they are not preprocessor files will run into this problem. There is already a unit test that verifies this: [ItOutputsFileWritesForProcessedContent](https://github.com/dotnet/sdk/blob/aca5d1556f7c9b3e2182a38844d66be663ff26fb/src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/GivenAProduceContentsAssetsTask.cs#L65)

In any case, this is moot because the fix to #577 will restrict the operation of this task to *.pp files because NuGet now handles other content files.

**Comment by natidea** [comment]
Closing as by design. We are using `ppOutputPath` to detect which content should be preprocessed in PR #580.

**Comment by nguerrera** [comment]
> Only content that incorrectly set the ppOutputPath metadata when they are not preprocessor files will run into this problem

@natidea What determines if ppOutputPath is set. There was definitely a package used by @harshjain2's team that had a dll that went through the preprocessor. Was this a nuget bug or a package authoring bug?

**Comment by natidea** [comment]
Fixed via #589. The ppOutputPath was being set to string.Empty instead of null.
